dace/codegen/targets/fpga.py

```diff
@@ -676,6 +851,162 @@ def deallocate_array(self, sdfg, dfg, state_id, node, function_stream,
+                            # We can stop here: this is a scope which will contain some compute, or a tasklet/libnode
+                            return True
+                    elif hasattr(curedge.src, "_kernel"):
+                        # Found a node with a kernel id. Use that
+                        return curedge.src._kernel
+                    next_edge = next(e for e in state.in_edges(curedge.src))
+                    curedge = next_edge
+
+                # We didn't return before
+                if not look_for_kernel_id:
+                    return False
+                else:
+                    return curedge.src._kernel if hasattr(
+                        curedge.src, "_kernel") else None
+
+            if hasattr(e.src, '_kernel'):
+                kernel = e.src._kernel
+
+                if (isinstance(e.dst, nodes.AccessNode)
+                        and isinstance(sdfg.arrays[e.dst.data], dt.View)):
+                    # Skip views
+                    e.dst._kernel = kernel
+                    continue
+
+                # Does this node need to be in another kernel?
+                # If it is a crossroad node (has more than one predecessor and its predecessors contain some compute)
+                # then it should be on a separate kernel.
+
+                if len(list(state.predecessors(e.dst))) > 1 and not isinstance(
+                        e.dst, nodes.ExitNode) and scopes[e.dst] == None:
+                    # Loop over all predecessors (except this edge)
+                    crossroad_node = False
+                    for pred_edge in state.in_edges(e.dst):
+                        if pred_edge != e and trace_back_edge(pred_edge, state):
+                            crossroad_node = True
+                            break
+
+                    if crossroad_node:
+                        kernel = max_kernels
+                        max_kernels = increment(max_kernels)
+
+            else:
+
+                # From this edge we don't have any kernel id.
+                # Look up for the other predecessor nodes, if any of them has a kernel
+                # ID, use that.
+
+                for pred_edge in state.in_edges(e.dst):
+                    if pred_edge != e:
+                        kernel = trace_back_edge(pred_edge,
+                                                 state,
+                                                 look_for_kernel_id=True)
+                        if kernel is not None:
+                            break
+                else:
+                    # Look at the successor nodes: because of the DFS visit, it may occur
+                    # that one of them has already an associated kernel ID. If this is the case, and
+                    # if the edge that connects this node with it is a tasklet-to-tasklet
+                    # edge, then we use that kernel ID. In all the other cases, we use a new one.
+
+                    # TODO: support more robust detection
```

Comment:
What are the current issues?

Reply:
There could be cases not covered by this. For example, I think that one should look also at the predecessors: if they are associated with a different kernel, and there is a tasklet-to-tasklet, maybe we don't want to generate a different kernel.
*Maybe* because I'm not sure...this could be further complicated as we want

Reply:
Put your concerns in the comment, currently it's not clear what is lacking in terms of robustness